src/Saunter/AsyncApiUiMiddleware.cs

```diff
@@ -0,0 +1,132 @@
+#if NETCOREAPP3_0
+using System;
+using System.Collections.Generic;
+using System.Net;
+using System.Net.Http;
+using System.Net.Mime;
+using System.Text.Json;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.Options;
+using Saunter.AsyncApiSchema.v2;
+using Saunter.Utils;
+
+namespace Saunter
+{
+    public class AsyncApiUiMiddleware
+    {
+
+        private const string GenerateApiPath = "html/generate";
+        private const string PlaygroundAssetsPath = "/html/template/";
+        private const string PlaygroundHtmlCacheKey = "index.html";
+
+        private readonly RequestDelegate _next;
+        private readonly IOptions<AsyncApiOptions> _options;
+        private readonly HttpClient _client;
+        private readonly IMemoryCache _memoryCache;
+
+        public AsyncApiUiMiddleware(RequestDelegate next, IOptions<AsyncApiOptions> options)
+        {
+            _next = next;
+            _options = options;
+            _client = new HttpClient
+            {
+                BaseAddress = new Uri(_options.Value.Middleware.PlaygroundBaseAddress)
+            };
+            _memoryCache = new MemoryCache(new MemoryCacheOptions());
+        }
+
+        public async Task Invoke(HttpContext context, IAsyncApiDocumentProvider asyncApiDocumentProvider)
+        {
+            if (!IsRequestingAsyncApiUi(context.Request) && !IsRequestingAsyncApiUiAssets(context.Request))
+            {
+                await _next(context);
+                return;
+            }
+
+            if (IsRequestingAsyncApiUi(context.Request))
+            {
+                var asyncApiSchema = asyncApiDocumentProvider.GetDocument();
+                await RespondWithAsyncApiHtml(context.Response, asyncApiSchema);
+                return;
+            }
+
+            await RespondWithProxiedResponse(context.Request, context.Response);
+        }
+
+        private async Task RespondWithAsyncApiHtml(HttpResponse response, AsyncApiDocument asyncApiSchema)
+        {
+            if (!_memoryCache.TryGetValue<string>(PlaygroundHtmlCacheKey, out var responseString))
+            {
+                var asyncApiSchemaJson = JsonSerializer.Serialize(
```

Comment:
Now we have 2 places which we need to get the serialized JSON (here and in AsyncApiMiddleware.cs), it would be a good idea to move this to the `IAsyncApiDocumentProvider` so we don't have to maintain this converter list in 2 places.

```csharp
public interface IAsyncApiDocumentProvider
{
    AsyncApiDocument GetDocument();

    string GetDocumentJson();
}
```

This `GetDocumentJson()` can call `GetDocument()` then `JsonSerializer.Serialize()` with the right options.